packages/govuk-frontend/src/govuk/common/index.jsdom.test.mjs

```diff
@@ -110,26 +110,198 @@ describe('Common JS utilities', () => {
+
+      const { dataset } = document.getElementById('app-example2')
+      const result = extractConfigByNamespace(
+        class Component {
+          /**
+           * @satisfies {Schema}
+           */
+          static schema = {
+            properties: {
+              i18n: { type: 'object' }
+            }
+          }
+        },
+        dataset,
+        'i18n'
+      )
+
+      expect(result).toEqual({ key1: 'One', key2: 'Two', key3: 'Three' })
+    })
+
+    it('can extract config from key-value pairs (with invalid namespace, last)', () => {
+      document.body.outerHTML = outdent`
+        <div id="app-example2"
+          data-i18n.key1="One"
+          data-i18n.key2="Two"
+          data-i18n.key3="Three"
+          data-i18n>
+        </div>
+      `
+
+      const { dataset } = document.getElementById('app-example2')
+      const result = extractConfigByNamespace(
+        class Component {
+          /**
+           * @satisfies {Schema}
+           */
+          static schema = {
+            properties: {
+              i18n: { type: 'object' }
+            }
+          }
+        },
+        dataset,
+        'i18n'
+      )
+
+      expect(result).toEqual({ key1: 'One', key2: 'Two', key3: 'Three' })
+    })
+
+    it('can handle multiple levels of nesting', () => {
+      document.body.outerHTML = outdent`
+        <div id="app-example2"
+          data-i18n.key1="This, That"
+          data-i18n.key2.one="The"
+          data-i18n.key2.other="Other">
+        </div>
+      `
+
+      const { dataset } = document.getElementById('app-example2')
+      const result = extractConfigByNamespace(
+        class Component {
+          /**
+           * @satisfies {Schema}
+           */
+          static schema = {
+            properties: {
+              i18n: { type: 'object' }
+            }
+          }
+        },
+        dataset,
+        'i18n'
+      )
+
+      expect(result).toEqual({
+        key1: 'This, That',
+        key2: {
+          one: 'The',
+          other: 'Other'
+        }
+      })
+    })
+
+    it('can handle multiple levels of nesting (prioritises the last parameter provided)', () => {
+      document.body.outerHTML = outdent`
+        <div id="app-example2"
+          data-i18n.key1.one="This"
+          data-i18n.key1.other="That"
+          data-i18n.key2.one="The"
+          data-i18n.key2.other="Other"
+          data-i18n.key1="This, That"
+          data-i18n.key2="The Other">
+        </div>
+      `
+
+      const { dataset } = document.getElementById('app-example2')
+      const result = extractConfigByNamespace(
+        class Component {
+          /**
+           * @satisfies {Schema}
+           */
+          static schema = {
+            properties: {
+              i18n: { type: 'object' }
+            }
+          }
+        },
+        dataset,
+        'i18n'
+      )
+
+      expect(result).toEqual({
+        key1: 'This, That',
+        key2: 'The Other'
+      })
+    })
```

Comment:
Thanks Romaric

You'll see the same [`prioritises the last parameter provided`](https://github.com/alphagov/govuk-frontend/blob/2c822ac4dc23fbdfb8d6370b0897c740ab51b90d/packages/govuk-frontend/src/govuk/common/index.jsdom.test.mjs#L75-L88) wording for `mergeConfigs()` too

To avoid surprises, I've made sure we keep the same merge order for `extractConfigByNamespace()`. When components list data attributes in a specific order, adding `.sort()` before looping would break that